docker_explorer/vhdx.py

```diff
@@ -0,0 +1,433 @@
+
+    Args:
+      bat_entry (BlockAllocationTableEntry): the BAT entry for the block
+        containing the target sector
+      sector_in_block (int): the target sector within the block
+
+    Returns:
+      bytes: either the sector contents or sector_size*b'\x00 if no offset was
+        given in the BAT entry
+    """
+    if bat_entry.offset:
+      self.vhdx_fd.seek(
+        bat_entry.offset + sector_in_block*self.logical_sector_size)
+      sector = self.vhdx_fd.read(self.logical_sector_size)
+    else:
+      sector = b'\x00'*self.logical_sector_size
+    return sector
+
+  def GetLogicalSector(self, block_number, sector_in_block):
+    """Returns a logical sector's contents
+
+    Args:
+      block_number (int): the block containing the sector
+      sector_in_block (int): the sector number within the target block
+
+    Returns:
+      bytes: the sector contents
+    """
+    bat_entry = self.bat_table.GetDBEntryForBlock(block_number)
+    state = bat_entry.state
+
+    if state == self.PAYLOAD_BLOCK_NOT_PRESENT:
+      if self.has_parent:
+        sector = self.parent_disk.GetLogicalSector(
+          block_number, sector_in_block)
+      else:
+        sector = self._GetLogicalSectorIfPresent(bat_entry, sector_in_block)
+    elif state == self.PAYLOAD_BLOCK_UNDEFINED:
+      sector = self._GetLogicalSectorIfPresent(bat_entry, sector_in_block)
+    elif state == self.PAYLOAD_BLOCK_ZERO:
+      sector = b'\x00'*self.logical_sector_size
+    elif state == self.PAYLOAD_BLOCK_UNMAPPED:
+      sector = self._GetLogicalSectorIfPresent(bat_entry, sector_in_block)
+    elif state == self.PAYLOAD_BLOCK_FULLY_PRESENT:
+      self.vhdx_fd.seek(
+        bat_entry.offset + sector_in_block*self.logical_sector_size)
+      sector = self.vhdx_fd.read(self.logical_sector_size)
+    elif state == self.PAYLOAD_BLOCK_PARTIALLY_PRESENT:
+      sector = self.parent_disk.GetLogicalSector(
+        block_number, sector_in_block)
+
+    return sector
+
+  def _GetDataBlockIfPresent(self, bat_entry):
+    """Returns block contents if an offset is present'
+
+    For certain block states the specification states that read behaviour can
+    either return the contents previously in the block or zeros.
+
+    Args:
+      bat_entry (BlockAllocationTableEntry): the BAT entry for the block
+        containing the target sector
+      sector_in_block (int): the target sector within the block
+
+    Returns:
+      bytes: either the block contents or block_size*b'\x00 if no offset was
+        given in the BAT entry
+    """
+    if bat_entry.offset:
+      self.vhdx_fd.seek(bat_entry.offset)
+      parsed_block = self.vhdx_fd.read(self.block_size)
+    else:
+      parsed_block = b'\x00'*self.block_size
+    return parsed_block
+
+  def GetDataBlock(self, block_number):
+    """Returns a data block's contents
+
+    Args:
+      block_number (int): the target block number
+
+    Returns:
+      bytes: the block contents
```

Comment:
Can you add Raises in the docstring?
This also applies to other functions like _GetSectorBitmapForBlock()